File: javascript/src/a2a.js
/**
 * A2A AgentCard signing and verification (v0.3.0).
 *
 * Mirrors the Rust `agentpin::a2a` module. AgentPin extends the
 * Google A2A AgentCard format with cryptographic identity verification.
 * The `agentpin` extension carries the AgentPin endpoint URL, the entity's
 * public key in JWK form, and a detached ECDSA P-256 signature over the
 * canonical bytes of the rest of the AgentCard.
 *
 * Canonicalisation: the signing input is the AgentCard with its `agentpin`
 * field cleared, serialised as JSON with object keys sorted alphabetically
 * (matches the Rust `serde_json::to_value` + `BTreeMap` trick).
 */

import { createPublicKey } from 'crypto';
import { signData, verifySignature } from './crypto.js';
import { pemToJwk, jwkToPem, jwkThumbprint } from './jwk.js';
import { AllowedDomains } from './discovery.js';
import { AgentPinError, ErrorCode } from './types.js';

/**
 * Map an AgentPin capability string (or `{ id }` Capability object) to a
 * minimal A2A AgentSkill.
 * @param {string|{id:string}} cap
 * @returns {{ id: string, name: string, description?: string }}
 */
export function capabilityToSkill(cap) {
    const id = typeof cap === 'string' ? cap : (cap && cap.id) || String(cap);
    return { id, name: id };
}

/**
 * Build an unsigned A2A AgentCard from an AgentPin `AgentDeclaration`.
 *
 * The capability list is mapped 1:1 to skills via `capabilityToSkill`; the
 * `allowed_domains` constraint is copied into `capabilities.allowed_domains`
 * (omitted entirely when unrestricted, matching the Rust serde behaviour).
 *
 * @param {string} url - Public URL where the agent receives A2A traffic.
 * @param {object} declaration - AgentPin AgentDeclaration.
 * @param {object} [opts]
 * @param {Array} [opts.skills] - Override the auto-mapped skill list.
 * @param {boolean} [opts.streaming]
 * @param {boolean} [opts.pushNotifications]
 * @returns {object} Unsigned A2A AgentCard (no `agentpin` extension).
 */
export function buildUnsignedAgentCard(url, declaration, opts = {}) {
    const skills = (opts.skills && opts.skills.length > 0)
        ? opts.skills.map((s) => ({ ...s }))
        : (declaration.capabilities || []).map(capabilityToSkill);

    const allowedDomains = declaration.constraints
        ? AllowedDomains.fromConstraints(declaration.constraints)
        : AllowedDomains.unrestricted();

    const capabilities = {
        streaming: !!opts.streaming,
        pushNotifications: !!opts.pushNotifications,
    };
    if (!AllowedDomains.isUnrestricted(allowedDomains)) {
        capabilities.allowed_domains = allowedDomains;
    }

    const card = {
        name: declaration.name,
        url,
        capabilities,
        skills,
    };
    if (declaration.description !== undefined && declaration.description !== null) {
        card.description = declaration.description;
    }
    if (declaration.version !== undefined && declaration.version !== null) {
        card.version = declaration.version;
    }
    return card;
}

/**
 * Sign an A2A AgentCard with an ECDSA P-256 private key.
 *
 * @param {object} unsignedCard - Card produced by `buildUnsignedAgentCard`.
 * @param {string} privateKeyPem
 * @param {string} kid
 * @param {string} agentpinEndpoint - URL of the entity's AgentPin discovery
 *   document (`.well-known/agent-identity.json`).
 * @returns {object} Signed A2A AgentCard with `agentpin` extension populated.
 */
export function signAgentCard(unsignedCard, privateKeyPem, kid, agentpinEndpoint) {
    if (!agentpinEndpoint) {
        throw new AgentPinError(
            ErrorCode.DISCOVERY_INVALID,
            'signAgentCard requires agentpinEndpoint'
        );
    }
    const canonical = canonicalizeForSigning({ ...unsignedCard, agentpin: undefined });
    const signature = signData(privateKeyPem, Buffer.from(canonical, 'utf8'));

    const publicKeyPem = createPublicKey(privateKeyPem)
        .export({ type: 'spki', format: 'pem' });
    const publicKeyJwk = pemToJwk(publicKeyPem, kid);

    return {
        ...unsignedCard,
        agentpin: {
            agentpin_endpoint: agentpinEndpoint,
            public_key_jwk: publicKeyJwk,
            signature,
        },
    };
}

/**
 * One-shot helper: build + sign in a single call.
 *
 * @param {string} url
 * @param {object} declaration
 * @param {string} privateKeyPem
 * @param {string} kid
 * @param {string} agentpinEndpoint
 * @param {object} [opts] - Forwarded to `buildUnsignedAgentCard`.
 * @returns {object} Signed A2A AgentCard.
 */
export function buildAndSignAgentCard(url, declaration, privateKeyPem, kid, agentpinEndpoint, opts = {}) {
    const unsigned = buildUnsignedAgentCard(url, declaration, opts);
    return signAgentCard(unsigned, privateKeyPem, kid, agentpinEndpoint);
}

/**
 * Verify the `agentpin` extension on an A2A AgentCard.
 *
 * Returns nothing on success; throws `AgentPinError(DISCOVERY_INVALID)` on
 * any failure (extension missing, malformed JWK, signature mismatch).
 *
 * This proves only that the card has not been tampered with relative to the
 * key inside its own extension. The caller still has to verify the JWK
 * chains back to a trusted AgentPin discovery document — pair this with
 * `A2aAgentCardResolver` for the full chain.
 *
 * @param {object} card - A2A AgentCard.
 */
export function verifyAgentpinExtension(card) {
    const ext = card && card.agentpin;
    if (!ext) {
        throw new AgentPinError(
            ErrorCode.DISCOVERY_INVALID,
            'AgentCard has no agentpin extension'
        );
    }
    const withoutExt = { ...card, agentpin: undefined };
    const canonical = canonicalizeForSigning(withoutExt);
    const publicKeyPem = jwkToPem(ext.public_key_jwk);
    const ok = verifySignature(publicKeyPem, Buffer.from(canonical, 'utf8'), ext.signature);
    if (!ok) {
        throw new AgentPinError(
            ErrorCode.DISCOVERY_INVALID,
            'A2A AgentCard signature did not verify against extension JWK'
        );
    }
}

/**
 * Compute the JWK thumbprint of the public key carried in a card's
 * `agentpin` extension. Convenience wrapper used by resolvers when matching
 * the card's key against a discovery document.
 *
 * @param {object} extension - Card's `agentpin` extension.
 * @returns {string} Hex thumbprint (no `sha256:` prefix).
 */
export function extensionKeyThumbprint(extension) {
    return jwkThumbprint(extension.public_key_jwk);
}

// ---------------------------------------------------------------------------
// Canonicalisation
// ---------------------------------------------------------------------------

/**
 * Canonicalise a JSON-able value with object keys sorted alphabetically and
 * `undefined` properties dropped — matches `serde_json::to_value` + sorted
 * `BTreeMap` re-serialisation in the Rust SDK so signatures verify across
 * languages.
 *
 * @param {*} value
 * @returns {string} Canonical JSON string.
 */
export function canonicalizeForSigning(value) {
    return JSON.stringify(sortedCanonical(value));
}

function sortedCanonical(value) {
    if (value === null || value === undefined) return null;
    if (Array.isArray(value)) return value.map(sortedCanonical);
    if (typeof value === 'object') {
        const out = {};
        for (const key of Object.keys(value).sort()) {
            const v = value[key];
            if (v === undefined) continue;
            out[key] = sortedCanonical(v);
        }
        return out;
    }
    return value;
}

File: python/agentpin/a2a.py
"""A2A AgentCard signing and verification (v0.3.0).

Mirrors the Rust ``agentpin::a2a`` module. AgentPin extends the Google A2A
AgentCard format with cryptographic identity verification. The ``agentpin``
extension carries the AgentPin endpoint URL, the entity's public key in JWK
form, and a detached ECDSA P-256 signature over the canonical bytes of the
rest of the AgentCard.

Canonicalisation: the signing input is the AgentCard with its ``agentpin``
field omitted, JSON-serialised with sorted keys and compact separators —
matches the Rust ``serde_json::to_value`` + ``BTreeMap`` trick.
"""

import json
from typing import Any, Dict, List, Optional, Union

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ec import EllipticCurvePrivateKey

from .crypto import sign_data, verify_signature
from .discovery import AllowedDomains
from .jwk import jwk_thumbprint, jwk_to_pem, pem_to_jwk
from .types import AgentPinError, ErrorCode


# ---------------------------------------------------------------------------
# Capability -> Skill mapping
# ---------------------------------------------------------------------------


def capability_to_skill(cap: Union[str, dict]) -> dict:
    """Map an AgentPin capability (string or ``{"id": ...}`` dict) to a skill."""
    if isinstance(cap, dict):
        id_ = cap.get("id", str(cap))
    else:
        id_ = str(cap)
    return {"id": id_, "name": id_}


# ---------------------------------------------------------------------------
# Builder
# ---------------------------------------------------------------------------


def build_unsigned_agent_card(
    url: str,
    declaration: dict,
    *,
    skills: Optional[List[dict]] = None,
    streaming: bool = False,
    push_notifications: bool = False,
) -> dict:
    """Build an unsigned A2A AgentCard from an AgentPin ``AgentDeclaration``.

    Capabilities map 1:1 to skills via ``capability_to_skill``; the
    ``allowed_domains`` constraint is copied into ``capabilities.allowed_domains``
    (omitted entirely when unrestricted, matching the Rust serde behaviour).
    """
    if skills is None or len(skills) == 0:
        out_skills = [capability_to_skill(c) for c in declaration.get("capabilities", [])]
    else:
        out_skills = [dict(s) for s in skills]

    allowed_domains = AllowedDomains.from_constraints(declaration.get("constraints"))

    capabilities: Dict[str, Any] = {
        "streaming": bool(streaming),
        "pushNotifications": bool(push_notifications),
    }
    if not AllowedDomains.is_unrestricted(allowed_domains):
        capabilities["allowed_domains"] = list(allowed_domains)

    card: Dict[str, Any] = {
        "name": declaration["name"],
        "url": url,
        "capabilities": capabilities,
        "skills": out_skills,
    }
    if declaration.get("description") is not None:
        card["description"] = declaration["description"]
    if declaration.get("version") is not None:
        card["version"] = declaration["version"]
    return card


def sign_agent_card(
    unsigned_card: dict,
    private_key_pem: str,
    kid: str,
    agentpin_endpoint: str,
) -> dict:
    """Sign an A2A AgentCard with an ECDSA P-256 private key (PEM).

    Returns the input card with the ``agentpin`` extension populated.
    """
    if not agentpin_endpoint:
        raise AgentPinError(
            ErrorCode.DISCOVERY_INVALID,
            "sign_agent_card requires agentpin_endpoint",
        )

    # Sign over the canonical bytes with the extension cleared.
    card_for_signing = {k: v for k, v in unsigned_card.items() if k != "agentpin"}
    canonical = canonicalize_for_signing(card_for_signing)
    signature = sign_data(private_key_pem, canonical.encode("utf-8"))

    # Derive the public-key JWK from the private key.
    private_key = serialization.load_pem_private_key(
        private_key_pem.encode("utf-8"), password=None
    )
    assert isinstance(private_key, EllipticCurvePrivateKey)
    public_pem = private_key.public_key().public_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    ).decode("utf-8")
    public_jwk = pem_to_jwk(public_pem, kid)

    signed = dict(unsigned_card)
    signed["agentpin"] = {
        "agentpin_endpoint": agentpin_endpoint,
        "public_key_jwk": public_jwk,
        "signature": signature,
    }
    return signed


def build_and_sign_agent_card(
    url: str,
    declaration: dict,
    private_key_pem: str,
    kid: str,
    agentpin_endpoint: str,
    *,
    skills: Optional[List[dict]] = None,
    streaming: bool = False,
    push_notifications: bool = False,
) -> dict:
    """One-shot helper: build + sign in a single call."""
    unsigned = build_unsigned_agent_card(
        url,
        declaration,
        skills=skills,
        streaming=streaming,
        push_notifications=push_notifications,
    )
    return sign_agent_card(unsigned, private_key_pem, kid, agentpin_endpoint)


# ---------------------------------------------------------------------------
# Verification
# ---------------------------------------------------------------------------


def verify_agentpin_extension(card: dict) -> None:
    """Verify the ``agentpin`` extension on an A2A AgentCard.

    Raises ``AgentPinError(DISCOVERY_INVALID)`` on any failure (extension
    missing, malformed JWK, signature mismatch).

    This proves only that the card has not been tampered with relative to the
    key inside its own extension. Pair with ``A2aAgentCardResolver`` for the
    full chain back to a trusted AgentPin discovery document.
    """
    ext = card.get("agentpin") if card else None
    if not ext:
        raise AgentPinError(
            ErrorCode.DISCOVERY_INVALID, "AgentCard has no agentpin extension"
        )

    without_ext = {k: v for k, v in card.items() if k != "agentpin"}
    canonical = canonicalize_for_signing(without_ext)
    public_pem = jwk_to_pem(ext["public_key_jwk"])
    ok = verify_signature(public_pem, canonical.encode("utf-8"), ext["signature"])
    if not ok:
        raise AgentPinError(
            ErrorCode.DISCOVERY_INVALID,
            "A2A AgentCard signature did not verify against extension JWK",
        )


def extension_key_thumbprint(extension: dict) -> str:
    """JWK thumbprint of the public key carried in a card's ``agentpin`` extension."""
    return jwk_thumbprint(extension["public_key_jwk"])


# ---------------------------------------------------------------------------
# Canonicalisation
# ---------------------------------------------------------------------------


def canonicalize_for_signing(value: Any) -> str:
    """Canonical JSON: sorted object keys, compact separators.

    Drops ``None`` values from objects so they round-trip identically with the
    Rust SDK's ``skip_serializing_if = "Option::is_none"`` behaviour.
    """
    return json.dumps(
        _sorted_canonical(value),
        sort_keys=False,
        separators=(",", ":"),
        ensure_ascii=False,
    )


def _sorted_canonical(value: Any) -> Any:
    if value is None:
        return None
    if isinstance(value, dict):
        out: Dict[str, Any] = {}
        for k in sorted(value.keys()):
            v = value[k]
            if v is None:
                continue
            out[k] = _sorted_canonical(v)
        return out
    if isinstance(value, list):
        return [_sorted_canonical(v) for v in value]
    return value
